server/src/v1/Tag/tag.service.ts

```
@@ -0,0 +1,72 @@
+import { StatusOk } from 'src/shared/Types'
+
+@Injectable()
+export class TagService {
+    private validator: Validator
+
+    constructor(
+        @InjectRepository(TagsRepository)
+        private readonly tagRepository: TagsRepository,
+        @InjectRepository(TitlesRepository)
+        private readonly titleRepository: TitlesRepository,
+    ) {
+        this.validator = new Validator()
+    }
+
+    async getTag(tagId: string): Promise<ISerializeResponse> {
+        if (!this.validator.isMongoId(tagId)) throw new BadRequestException('Id must be a type of MongoId')
+
+        let tag
+        try {
+            tag = await this.tagRepository.findOneOrFail(tagId)
+        }
+        catch {
+            throw new NotFoundException('Tag could not found by given id')
+        }
+
+        return serializerService.serializeResponse('tag_detail', tag)
+    }
+
+    async searchTag(searchValue: string): Promise<ISerializeResponse> {
+        if (searchValue.length < 3) throw new BadRequestException('Search value must be greater than 2 characters')
+        const result = await this.tagRepository.searchTag(searchValue)
+        return serializerService.serializeResponse('tag_search_result', result)
+    }
+
+    async getTrendingTags(): Promise<ISerializeResponse> {
+        const result: {
+            tags: TagsEntity[],
+            count: number
+        } = await this.tagRepository.getTrendingTags()
+        return serializerService.serializeResponse('trending_tag_list', result)
+    }
+
+    async deleteTag(tagId: string): Promise<StatusOk> {
+        if (!this.validator.isMongoId(tagId)) throw new BadRequestException('Id must be a type of MongoId')
+
+        let tag: TagsEntity | null
+        try {
+            tag = await this.tagRepository.findOneOrFail(tagId)
+        }
+        catch {
+            throw new NotFoundException('Tag could not found by given id')
+        }
+
+        await this.titleRepository.deleteTagFromTitle(tag.name)
```

We should wrap this in a transaction block.

We want to ensure that we `deleteTagFromTitle` AND `deleteTag`.

If either one fails, we want to the db to roll back.

think of it like "all or nothing".

Imagine what happens to our data integrity if `deleteTag` fails silently.

Titles will not have tags, but the tags still exist in the db.